src/utils/thread_helper.cpp

```diff
@@ -199,10 +199,10 @@ void ThreadHelper::EnergyScanCallback(otEnergyScanResult *aResult)
 {
     if (aResult == nullptr)
     {
-        if (mEnergyScanHandler != nullptr)
-        {
-            mEnergyScanHandler(OT_ERROR_NONE, mEnergyScanResults);
-        }
+        assert(mEnergyScanHandler != nullptr);
+
+        mEnergyScanHandler(OT_ERROR_NONE, mEnergyScanResults);
+        mEnergyScanHandler = nullptr;
```

Comment:
We should be extremely careful when calling callbacks. The callback may be reset during the handler. 

```suggestion
        auto callback = mEnergyScanHandler;
        mEnergyScanHandler = nullptr;
        callback(OT_ERROR_NONE, mEnergyScanResults);
```